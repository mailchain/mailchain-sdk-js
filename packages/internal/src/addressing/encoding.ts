import { EncodingTypes, EncodingType } from '@mailchain/encoding';
import { Decode, Encode } from '@mailchain/encoding/encoding';
import { ALGORAND, ALL_PROTOCOLS, ETHEREUM, ProtocolType, SUBSTRATE } from '../protocols';

/**
 * Convert address from Uint8Array, selects the relevant encoding method and encodes it as string.
 */
export function encodeAddressByProtocol(
	address: Uint8Array,
	protocol: ProtocolType,
): { encoded: string; encoding: string } {
	const encoding = encodingByProtocol(protocol);
	const encoded = Encode(encoding, address);

	return { encoded, encoding };
}

/**
 * Convert address from string, selects the relevant encoding method and decodes it as Uint8Array.
 */
export function decodeAddressByProtocol(
	address: string,
	protocol: ProtocolType,
): { decoded: Uint8Array; encoding: string } {
	const encoding = encodingByProtocol(protocol);
	const decoded = Decode(encoding, address);

	return { decoded, encoding };
}

/**
 * EncodingByProtocol returns the relevant encoding method the protocol commonly uses.
 */
//
export function encodingByProtocol(protocol: ProtocolType): EncodingType {
	switch (protocol) {
		case ALGORAND:
			return EncodingTypes.Base32;
		case ETHEREUM:
			return EncodingTypes.Hex0xPrefix;
		case SUBSTRATE:
			return EncodingTypes.Base58;
		default:
			throw new Error(`unknown address encoding of [${protocol}]. Supported ${ALL_PROTOCOLS}.`);
	}
}
