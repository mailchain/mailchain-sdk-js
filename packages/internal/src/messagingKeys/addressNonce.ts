import {
	ETHEREUM,
	MAILCHAIN,
	NEAR,
	ProtocolNotSupportedError,
	ProtocolType,
	TEZOS,
	isBlockchainProtocolEnabled,
} from '@mailchain/addressing';
import axios, { AxiosInstance } from 'axios';
import { MessagingKeysApiFactory, MessagingKeysApiInterface, createAxiosConfiguration } from '@mailchain/api';
import { Configuration } from '../configuration';
import { MailchainResult } from '../mailchainResult';
import { ContractCallLatestNonce } from './contractResolvers/resolver';
import { NearContractCallResolver } from './contractResolvers/near';
import { MailchainKeyRegContractCallResolver } from './contractResolvers/mailchain';
import { MessagingKeyNotFoundInContractError } from './contractResolvers/errors';
import { UnexpectedMailchainError } from './errors';

export type GetMessagingKeyLatestNonceResult = MailchainResult<number, GetMessagingKeyLatestNonceError>;
export type GetMessagingKeyLatestNonceError = ProtocolNotSupportedError | UnexpectedMailchainError;

export class AddressNonce {
	constructor(
		private readonly messagingKeysApi: MessagingKeysApiInterface,
		private readonly resolvers: Map<ProtocolType, ContractCallLatestNonce>,
	) {}

	static create(configuration: Configuration, axiosInstance: AxiosInstance = axios.create()) {
		const mailchainKeyRegistryResolver = MailchainKeyRegContractCallResolver.create(configuration, axiosInstance);
		return new AddressNonce(
			MessagingKeysApiFactory(createAxiosConfiguration(configuration.apiPath)),
			new Map<ProtocolType, ContractCallLatestNonce>([
				[NEAR, NearContractCallResolver.create(configuration, axiosInstance)],
				[ETHEREUM, mailchainKeyRegistryResolver],
				[TEZOS, mailchainKeyRegistryResolver],
			]),
		);
	}

	/**
	 * Get the latest nonce for an address.
	 *
	 * @param address the protocol get the latest nonce for.
	 * @param protocol where to find the address.
	 * @returns The latest nonce for the given address.
	 */
	async getMessagingKeyLatestNonce(
		address: string,
		protocol: ProtocolType,
	): Promise<GetMessagingKeyLatestNonceResult> {
		if (protocol === MAILCHAIN) {
			return { data: 1 }; // currently mailchain accounts do not support incrementing nonces
		}

		const resolver = this.resolvers.get(protocol);
		if (!resolver) {
			return { error: new ProtocolNotSupportedError(protocol) };
		}

		if (!isBlockchainProtocolEnabled(protocol)) {
			return { error: new ProtocolNotSupportedError(protocol) };
		}

		const nonceContractResponse = await this.messagingKeysApi.getProtocolAddressNonce(address, protocol);

		try {
			const nonce = await resolver.latestNonce(nonceContractResponse.data.contractCall);
			return {
				data: nonce,
			};
		} catch (error) {
			if (error instanceof MessagingKeyNotFoundInContractError) {
				return {
					data: 0,
				};
			}
			return {
				error: new UnexpectedMailchainError('Failed to get latest nonce.', error as Error),
			};
		}
	}
}
