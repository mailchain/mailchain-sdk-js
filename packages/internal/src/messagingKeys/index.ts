export * from './addressNonce';
export * from './errors';
export * from './messagingKeys';
export * from './messagingKeys';
export * from './privateMessagingKeys';
export * from './proof';
