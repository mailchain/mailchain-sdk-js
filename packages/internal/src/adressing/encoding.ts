import { EncodingTypes, EncodingType } from '@mailchain/encoding';
import { Encode } from '@mailchain/encoding/encoding';
import { Algorand, Ethereum, Substrate } from '../protocols';

/**
 * EncodeAddressByProtocol takes an address as an Uint8Array then selects the relevant encoding method to encode it as string.
 * @param address
 * @param param1
 * @param byte
 * @param protocol
 * @param string
 */
export function EncodeAddressByProtocol(address: Uint8Array, protocol: string): { encoded: string; encoding: string } {
	const encoding = EncodingByProtocol(protocol);
	const encoded = Encode(encoding, address);

	return { encoded, encoding };
}

/**
 * EncodingByProtocol returns the relevant encoding method the protocol commonly uses.
 */
//
export function EncodingByProtocol(protocol: string): EncodingType {
	switch (protocol) {
		case Algorand:
			return EncodingTypes.Base32;
		case Ethereum:
			return EncodingTypes.Hex0xPrefix;
		case Substrate:
			return EncodingTypes.Base58;
		default:
			throw new Error('unknown address encoding');
	}
}
