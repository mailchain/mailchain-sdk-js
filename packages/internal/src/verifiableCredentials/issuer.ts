import { SignerWithPublicKey, isPublicKeyEqual } from '@mailchain/crypto';
import { decodeUtf8, encodeBase64UrlSafe } from '@mailchain/encoding';
import { Issuer } from 'did-jwt-vc';
import { ProvidedMessagingKeyIncorrectError } from '@mailchain/signatures';
import { UnexpectedMailchainError } from '../errors';
import { MailchainResult } from '../mailchainResult';
import { RegisteredResolvedAddress } from '../messagingKeys';
import { mailchainAddressDecentralizedIdentifier } from './did';

export type CreateMailchainMessagingKeyIssuerError = UnexpectedMailchainError | ProvidedMessagingKeyIncorrectError;
export type CreateIssuerFromResolvedAddress = {
	/**
	 * Resolved address signer will be created for. Resolved address public key must match signer public key.
	 */
	resolvedAddress: RegisteredResolvedAddress;
	/**
	 * Signer must match resolved address public key. Used to sign issued credentials.
	 */
	signer: SignerWithPublicKey;
};

export class MailchainMessagingKeyIssuer {
	async createIssuerFromResolvedAddress(
		params: CreateIssuerFromResolvedAddress,
	): Promise<MailchainResult<Issuer, CreateMailchainMessagingKeyIssuerError>> {
		const { resolvedAddress, signer } = params;

		if (!isPublicKeyEqual(resolvedAddress.messagingKey, signer.publicKey)) {
			return {
				error: new ProvidedMessagingKeyIncorrectError('signer'),
			};
		}

		return {
			data: createIssuerFromSigner(resolvedAddress.mailchainAddress, signer),
		};
	}
}

export function createIssuerFromSigner(mailchainAddress: string, signer: SignerWithPublicKey): Issuer {
	return {
		did: mailchainAddressDecentralizedIdentifier(mailchainAddress),
		alg: 'Ed25519',
		signer: async (data: string | Uint8Array): Promise<string> => {
			const dataBytes: Uint8Array = typeof data === 'string' ? decodeUtf8(data) : data;
			const signature = await signer.sign(dataBytes);
			return encodeBase64UrlSafe(signature);
		},
	};
}
