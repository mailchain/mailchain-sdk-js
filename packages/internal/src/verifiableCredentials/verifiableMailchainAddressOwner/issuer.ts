import { VerifiableCredential, createVerifiablePresentationJwt } from 'did-jwt-vc';
import { SignerWithPublicKey, publicKeyToBytes } from '@mailchain/crypto';
import { Proof } from 'did-jwt-vc/lib/types';
import { encodeHex } from '@mailchain/encoding';
import {
	AddressNotRegisteredError,
	MessagingKeys,
	RegisteredResolvedAddress,
	ResolveAddressError,
} from '../../messagingKeys';
import { Configuration, MailchainResult } from '../..';
import {
	DecentralizedIdentifier,
	mailchainAddressDecentralizedIdentifier,
	mailchainBlockchainAddressDecentralizedIdentifier,
} from '../did';
import { createPresentationPayload } from '../presentation';
import { CreateMailchainMessagingKeyIssuerError, MailchainMessagingKeyIssuer } from '../issuer';
import { VerifiablePresentationJWT } from '../jwt';
import { createVerifiableCredential } from '../payload';
import { createOwnerOfMessagingKeySubject } from '../subject';

export type CreateVerifiableMailchainAddressOwnerError =
	| CreateMailchainMessagingKeyIssuerError
	| AddressNotRegisteredError
	| ResolveAddressError;

export type CreateVerifiableMailchainAddressOwnerParams = {
	/**
	 * Uniquely identify request.
	 */
	id?: string;
	/**
	 * Recipient Mailchain address.
	 */
	address: string;
	signer: SignerWithPublicKey;
	/**
	 * Who is requesting the verifiable credential.
	 */
	requester: string;
	/**
	 * The actions that the holder is allowed to perform.
	 * @example ['Authenticate', 'Join meeting']
	 */
	actions: string[];

	resources: string[];
	options: {
		/**
		 * The number of seconds from the current time that the credential will expire.
		 */
		expiresIn?: number;
		/**
		 * The date and time that the credential will expire.
		 */
		expiresAt?: Date;
		nonce?: string;
	};
};

function createIssuerProof(resolvedAddress: RegisteredResolvedAddress): Proof {
	switch (resolvedAddress.proof.source) {
		case 'MailchainRegistry':
			const { signature, address, messagingKey, identityKey, ...remainingProof } = resolvedAddress.proof;
			return {
				signatureHex: encodeHex(signature),
				addressHex: encodeHex(address),
				messagingKeyHex: encodeHex(publicKeyToBytes(messagingKey)),
				identityKeyHex: encodeHex(publicKeyToBytes(identityKey)),
				...remainingProof,
				type: 'Mailchain-Ed25519Signature2020',
			};
		case 'ContractCall':
			return {
				...resolvedAddress.proof,
				type: 'Mailchain-Ed25519Signature2020',
			};
		default:
			throw new Error(`unknown proof source ${(resolvedAddress.proof as Proof).source}`);
	}
}

function createCredentialPayloadMailchainAddressOwner(
	resolvedAddress: RegisteredResolvedAddress,
	requester: DecentralizedIdentifier<string>,
	actions: string[],
	resources: string[],
): VerifiableCredential {
	return createVerifiableCredential({
		type: 'MailchainMessagingKeyCredential',
		credentialSubjects: [createOwnerOfMessagingKeySubject(resolvedAddress.mailchainAddress)],
		issuanceDate: new Date(),
		issuerId: mailchainBlockchainAddressDecentralizedIdentifier(
			resolvedAddress.protocol,
			resolvedAddress.protocolAddress,
		),
		proof: createIssuerProof(resolvedAddress),
		termsOfUse: [
			{
				type: 'HolderPolicy',
				assigner: mailchainAddressDecentralizedIdentifier(resolvedAddress.mailchainAddress),
				effect: 'Allow',
				assignee: requester,
				actions,
				resources,
			},
		],
	});
}

export class MailchainAddressOwnershipIssuer {
	constructor(
		private readonly messagingKeys: MessagingKeys,
		private readonly mailchainMessagingKeyIssuer: MailchainMessagingKeyIssuer,
	) {}

	static create(configuration: Configuration) {
		return new MailchainAddressOwnershipIssuer(
			MessagingKeys.create(configuration),
			new MailchainMessagingKeyIssuer(),
		);
	}

	async createVerifiableMailchainAddressOwnership(
		params: CreateVerifiableMailchainAddressOwnerParams,
	): Promise<MailchainResult<VerifiablePresentationJWT, CreateVerifiableMailchainAddressOwnerError>> {
		const { address, signer, requester, options, actions, resources, id } = params;
		const { data: resolvedAddress, error: resolveAddressError } = await this.messagingKeys.resolve(address);
		if (resolveAddressError) {
			return {
				error: resolveAddressError,
			};
		}

		if (resolvedAddress.type !== 'registered') {
			return {
				error: new AddressNotRegisteredError(),
			};
		}

		const { expiresIn, expiresAt, nonce } = options;

		const presentationPayload = createPresentationPayload({
			id,
			verifiableCredential: createCredentialPayloadMailchainAddressOwner(
				resolvedAddress,
				mailchainAddressDecentralizedIdentifier(requester),
				actions,
				resources,
			),
			holder: mailchainAddressDecentralizedIdentifier(resolvedAddress.mailchainAddress),
			verifier: mailchainAddressDecentralizedIdentifier(requester),
			expiresAt,
		});

		const { data: issuer, error: createMailchainMessagingKeyIssuerError } =
			await this.mailchainMessagingKeyIssuer.createIssuerFromResolvedAddress({
				resolvedAddress,
				signer,
			});

		if (createMailchainMessagingKeyIssuerError) {
			return {
				error: createMailchainMessagingKeyIssuerError,
			};
		}

		const verifiablePresentation = await createVerifiablePresentationJwt(presentationPayload, issuer, {
			// challenge and domain are used to prevent replay attacks
			challenge: nonce,
			domain: mailchainAddressDecentralizedIdentifier(requester),
			canonicalize: true,
			expiresIn,
		});

		return { data: verifiablePresentation };
	}
}
