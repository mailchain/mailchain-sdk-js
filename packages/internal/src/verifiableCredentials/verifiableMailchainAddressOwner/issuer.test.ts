import { mock } from 'jest-mock-extended';
import {
	AliceED25519PrivateKey,
	AliceED25519PublicKey,
	BobED25519PrivateKey,
	BobED25519PublicKey,
} from '@mailchain/crypto/ed25519/test.const';
import {
	AliceSECP256K1PublicAddress,
	AliceSECP256K1PublicAddressStr,
	BobSECP256K1PublicAddress,
	BobSECP256K1PublicAddressStr,
} from '@mailchain/addressing/protocols/ethereum/test.const';
import { AliceSECP256K1PublicKey, BobSECP256K1PublicKey } from '@mailchain/crypto/secp256k1/test.const';
import { decodeBase64, encodeUtf8 } from '@mailchain/encoding';
import { MessagingKeys } from '../../messagingKeys';
import { MailchainMessagingKeyIssuer, createIssuerFromSigner } from '../issuer';
import { MailchainAddressOwnershipIssuer } from './issuer';

const mockMessagingKeys = mock<MessagingKeys>();
mockMessagingKeys.resolve.calledWith(`${AliceSECP256K1PublicAddressStr}@ethereum.mailchain.com`).mockResolvedValue({
	data: {
		mailchainAddress: `${AliceSECP256K1PublicAddressStr}@ethereum.mailchain.com`,
		messagingKey: AliceED25519PublicKey,
		protocol: 'ethereum',
		protocolAddress: AliceSECP256K1PublicAddressStr,
		type: 'registered',
		identityKey: AliceSECP256K1PublicKey,
		proof: {
			address: AliceSECP256K1PublicAddress,
			identityKey: AliceSECP256K1PublicKey,
			locale: 'locale',
			messageVariant: 'v1',
			messagingKey: AliceED25519PublicKey,
			nonce: 1,
			network: '',
			protocol: 'ethereum',
			signature: Uint8Array.from([1, 2, 3, 4]),
			signatureMethod: 'ethereum_personal_message',
			source: 'MailchainRegistry',
		},
	},
});
mockMessagingKeys.resolve.calledWith(`${BobSECP256K1PublicAddressStr}@ethereum.mailchain.com`).mockResolvedValue({
	data: {
		mailchainAddress: `${BobSECP256K1PublicAddressStr}@ethereum.mailchain.com`,
		messagingKey: BobED25519PublicKey,
		protocol: 'ethereum',
		protocolAddress: BobSECP256K1PublicAddressStr,
		type: 'registered',
		identityKey: BobSECP256K1PublicKey,
		proof: {
			address: BobSECP256K1PublicAddress,
			identityKey: BobSECP256K1PublicKey,
			locale: 'locale',
			messageVariant: 'v1',
			messagingKey: BobED25519PublicKey,
			nonce: 1,
			network: '',
			protocol: 'ethereum',
			signature: Uint8Array.from([1, 2, 3, 4]),
			signatureMethod: 'ethereum_personal_message',
			source: 'MailchainRegistry',
		},
	},
});

const mockMailchainMessagingKeyIssuer = mock<MailchainMessagingKeyIssuer>();

describe('issuer', () => {
	beforeAll(() => {
		jest.useFakeTimers().setSystemTime(new Date(2020, 1, 2));
	});

	let issuer: MailchainAddressOwnershipIssuer;

	beforeEach(() => {
		jest.clearAllMocks();

		issuer = new MailchainAddressOwnershipIssuer(mockMessagingKeys, mockMailchainMessagingKeyIssuer);
	});
	it('mailchain-address-alice', async () => {
		mockMailchainMessagingKeyIssuer.createIssuerFromResolvedAddress.mockResolvedValue({
			data: createIssuerFromSigner(
				`${AliceSECP256K1PublicAddressStr}@ethereum.mailchain.com`,
				AliceED25519PrivateKey,
			),
		});

		const result = await issuer.createVerifiableMailchainAddressOwnership({
			address: `${AliceSECP256K1PublicAddressStr}@ethereum.mailchain.com`,
			signer: AliceED25519PrivateKey,
			options: {
				expiresIn: 600,
				nonce: '1234',
			},
			actions: ['Authenticate', 'Join Meeting'],
			resources: ['*'],
			requester: 'app-example@mailchain.com',
		});

		expect(result.error).toBeUndefined();
		expect(result.data).toMatchSnapshot();
		expect(encodeUtf8(decodeBase64(result.data!.split('.')[1]))).toMatchSnapshot();
	});

	it('mailchain-address-bob', async () => {
		mockMailchainMessagingKeyIssuer.createIssuerFromResolvedAddress.mockResolvedValue({
			data: createIssuerFromSigner(
				`${BobSECP256K1PublicAddressStr}@ethereum.mailchain.com`,
				BobED25519PrivateKey,
			),
		});

		const result = await issuer.createVerifiableMailchainAddressOwnership({
			address: `${BobSECP256K1PublicAddressStr}@ethereum.mailchain.com`,
			signer: BobED25519PrivateKey,
			options: {
				expiresIn: 600,
				nonce: '1234',
			},
			actions: ['Authenticate', 'Join Meeting'],
			resources: ['*'],
			requester: 'app-example@mailchain.com',
		});

		expect(result.error).toBeUndefined();
		expect(result.data).toMatchSnapshot();
		expect(encodeUtf8(decodeBase64(result.data!.split('.')[1]))).toMatchSnapshot();
	});
});
