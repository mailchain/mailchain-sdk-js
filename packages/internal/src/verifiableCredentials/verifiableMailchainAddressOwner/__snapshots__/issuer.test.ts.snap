// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issuer mailchain-address-alice 1`] = `"eyJhbGciOiJFZDI1NTE5IiwidHlwIjoiSldUIn0.eyJhdWQiOlsiZGlkOm1haWxjaGFpbjphcHAtZXhhbXBsZSU0MG1haWxjaGFpbi5jb20iXSwiZXhwIjoxNTgwNjAyMjAwLCJpc3MiOiJkaWQ6bWFpbGNoYWluOjB4RDVhYjRDRTM2MDVDZDU5MERiNjA5YjZiNUM4OTAxZmRCMmVmN0ZFNiU0MGV0aGVyZXVtLm1haWxjaGFpbi5jb20iLCJuYmYiOjE1ODA2MDE2MDAsIm5vbmNlIjoiMTIzNCIsInZwIjp7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIl0sInR5cGUiOlsiVmVyaWZpYWJsZVByZXNlbnRhdGlvbiJdLCJ2ZXJpZmlhYmxlQ3JlZGVudGlhbCI6W3siQGNvbnRleHQiOlsiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvdjEiXSwiY3JlZGVudGlhbFN1YmplY3QiOlt7Im93bmVyT2YiOnsiYWRkcmVzcyI6IjB4RDVhYjRDRTM2MDVDZDU5MERiNjA5YjZiNUM4OTAxZmRCMmVmN0ZFNkBldGhlcmV1bS5tYWlsY2hhaW4uY29tIiwidHlwZSI6Ik1haWxjaGFpbk1lc3NhZ2luZ0tleSJ9fV0sImlzc3VhbmNlRGF0ZSI6IjIwMjAtMDItMDJUMDA6MDA6MDAuMDAwWiIsImlzc3VlciI6eyJpZCI6ImRpZDptYWlsY2hhaW46ZXRoZXJldW06MHhENWFiNENFMzYwNUNkNTkwRGI2MDliNmI1Qzg5MDFmZEIyZWY3RkU2In0sInByb29mIjp7ImFkZHJlc3NIZXgiOiJkNWFiNGNlMzYwNWNkNTkwZGI2MDliNmI1Yzg5MDFmZGIyZWY3ZmU2IiwiaWRlbnRpdHlLZXlIZXgiOiJlMTAyNjlkOTA4NTEwZTM1NWJlYjFkNWJmMmRmODEyOWU1YjY0MDFlMTk2OTg5MWU4MDE2YTBiMjMwMDczOWJiYjAwNiIsImxvY2FsZSI6ImxvY2FsZSIsIm1lc3NhZ2VWYXJpYW50IjoidjEiLCJtZXNzYWdpbmdLZXlIZXgiOiJlMjcyM2NhYTIzYTViNTExYWY1YWQ3YjdlZjYwNzZlNDE0YWI3ZTc1YTlkYzkxMGVhNjBlNDE3YTJiNzcwYTU2NzEiLCJuZXR3b3JrIjoiIiwibm9uY2UiOjEsInByb3RvY29sIjoiZXRoZXJldW0iLCJzaWduYXR1cmVIZXgiOiIwMTAyMDMwNCIsInNpZ25hdHVyZU1ldGhvZCI6ImV0aGVyZXVtX3BlcnNvbmFsX21lc3NhZ2UiLCJzb3VyY2UiOiJNYWlsY2hhaW5SZWdpc3RyeSIsInR5cGUiOiJNYWlsY2hhaW4tRWQyNTUxOVNpZ25hdHVyZTIwMjAifSwidGVybXNPZlVzZSI6W3siYWN0aW9ucyI6WyJBdXRoZW50aWNhdGUiLCJKb2luIE1lZXRpbmciXSwiYXNzaWduZWUiOiJkaWQ6bWFpbGNoYWluOmFwcC1leGFtcGxlJTQwbWFpbGNoYWluLmNvbSIsImFzc2lnbmVyIjoiZGlkOm1haWxjaGFpbjoweEQ1YWI0Q0UzNjA1Q2Q1OTBEYjYwOWI2YjVDODkwMWZkQjJlZjdGRTYlNDBldGhlcmV1bS5tYWlsY2hhaW4uY29tIiwiZWZmZWN0IjoiQWxsb3ciLCJyZXNvdXJjZXMiOlsiKiJdLCJ0eXBlIjoiSG9sZGVyUG9saWN5In1dLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiTWFpbGNoYWluTWVzc2FnaW5nS2V5Q3JlZGVudGlhbCJdfV19fQ.o7gpQmN0WAE1yeSNR5xT0YnXfuC5E5tvfFVGipHV8FwRdkvAzKG5OOWRc8jRCsn5QT4Qu9bl55OGkuxjmNoHCQ"`;

exports[`issuer mailchain-address-alice 2`] = `"{"aud":["did:mailchain:app-example%40mailchain.com"],"exp":1580602200,"iss":"did:mailchain:0xD5ab4CE3605Cd590Db609b6b5C8901fdB2ef7FE6%40ethereum.mailchain.com","nbf":1580601600,"nonce":"1234","vp":{"@context":["https://www.w3.org/2018/credentials/v1"],"type":["VerifiablePresentation"],"verifiableCredential":[{"@context":["https://www.w3.org/2018/credentials/v1"],"credentialSubject":[{"ownerOf":{"address":"0xD5ab4CE3605Cd590Db609b6b5C8901fdB2ef7FE6@ethereum.mailchain.com","type":"MailchainMessagingKey"}}],"issuanceDate":"2020-02-02T00:00:00.000Z","issuer":{"id":"did:mailchain:ethereum:0xD5ab4CE3605Cd590Db609b6b5C8901fdB2ef7FE6"},"proof":{"addressHex":"d5ab4ce3605cd590db609b6b5c8901fdb2ef7fe6","identityKeyHex":"e10269d908510e355beb1d5bf2df8129e5b6401e1969891e8016a0b2300739bbb006","locale":"locale","messageVariant":"v1","messagingKeyHex":"e2723caa23a5b511af5ad7b7ef6076e414ab7e75a9dc910ea60e417a2b770a5671","network":"","nonce":1,"protocol":"ethereum","signatureHex":"01020304","signatureMethod":"ethereum_personal_message","source":"MailchainRegistry","type":"Mailchain-Ed25519Signature2020"},"termsOfUse":[{"actions":["Authenticate","Join Meeting"],"assignee":"did:mailchain:app-example%40mailchain.com","assigner":"did:mailchain:0xD5ab4CE3605Cd590Db609b6b5C8901fdB2ef7FE6%40ethereum.mailchain.com","effect":"Allow","resources":["*"],"type":"HolderPolicy"}],"type":["VerifiableCredential","MailchainMessagingKeyCredential"]}]}}"`;

exports[`issuer mailchain-address-bob 1`] = `"eyJhbGciOiJFZDI1NTE5IiwidHlwIjoiSldUIn0.eyJhdWQiOlsiZGlkOm1haWxjaGFpbjphcHAtZXhhbXBsZSU0MG1haWxjaGFpbi5jb20iXSwiZXhwIjoxNTgwNjAyMjAwLCJpc3MiOiJkaWQ6bWFpbGNoYWluOjB4OTJkOGYxMDI0OGM2YTM5NTNjYzM2OTJhODk0NjU1YWQwNWQ2MWVmYiU0MGV0aGVyZXVtLm1haWxjaGFpbi5jb20iLCJuYmYiOjE1ODA2MDE2MDAsIm5vbmNlIjoiMTIzNCIsInZwIjp7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIl0sInR5cGUiOlsiVmVyaWZpYWJsZVByZXNlbnRhdGlvbiJdLCJ2ZXJpZmlhYmxlQ3JlZGVudGlhbCI6W3siQGNvbnRleHQiOlsiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvdjEiXSwiY3JlZGVudGlhbFN1YmplY3QiOlt7Im93bmVyT2YiOnsiYWRkcmVzcyI6IjB4OTJkOGYxMDI0OGM2YTM5NTNjYzM2OTJhODk0NjU1YWQwNWQ2MWVmYkBldGhlcmV1bS5tYWlsY2hhaW4uY29tIiwidHlwZSI6Ik1haWxjaGFpbk1lc3NhZ2luZ0tleSJ9fV0sImlzc3VhbmNlRGF0ZSI6IjIwMjAtMDItMDJUMDA6MDA6MDAuMDAwWiIsImlzc3VlciI6eyJpZCI6ImRpZDptYWlsY2hhaW46ZXRoZXJldW06MHg5MmQ4ZjEwMjQ4YzZhMzk1M2NjMzY5MmE4OTQ2NTVhZDA1ZDYxZWZiIn0sInByb29mIjp7ImFkZHJlc3NIZXgiOiI5MmQ4ZjEwMjQ4YzZhMzk1M2NjMzY5MmE4OTQ2NTVhZDA1ZDYxZWZiIiwiaWRlbnRpdHlLZXlIZXgiOiJlMTAzYmRmNmZiOTdjOTdjMTI2YjQ5MjE4NmE0ZDViMjhmMzRmMDY3MWE1YWFjYzk3NGRhM2JkZTBiZTkzZTQ1YTFjNSIsImxvY2FsZSI6ImxvY2FsZSIsIm1lc3NhZ2VWYXJpYW50IjoidjEiLCJtZXNzYWdpbmdLZXlIZXgiOiJlMjJlMzIyZjg3NDBjNjAxNzIxMTFhYzhlYWRjZGRhMjUxMmY5MGQwNmQwZTUwM2VmMTg5OTc5YTE1OWJlY2UxZTgiLCJuZXR3b3JrIjoiIiwibm9uY2UiOjEsInByb3RvY29sIjoiZXRoZXJldW0iLCJzaWduYXR1cmVIZXgiOiIwMTAyMDMwNCIsInNpZ25hdHVyZU1ldGhvZCI6ImV0aGVyZXVtX3BlcnNvbmFsX21lc3NhZ2UiLCJzb3VyY2UiOiJNYWlsY2hhaW5SZWdpc3RyeSIsInR5cGUiOiJNYWlsY2hhaW4tRWQyNTUxOVNpZ25hdHVyZTIwMjAifSwidGVybXNPZlVzZSI6W3siYWN0aW9ucyI6WyJBdXRoZW50aWNhdGUiLCJKb2luIE1lZXRpbmciXSwiYXNzaWduZWUiOiJkaWQ6bWFpbGNoYWluOmFwcC1leGFtcGxlJTQwbWFpbGNoYWluLmNvbSIsImFzc2lnbmVyIjoiZGlkOm1haWxjaGFpbjoweDkyZDhmMTAyNDhjNmEzOTUzY2MzNjkyYTg5NDY1NWFkMDVkNjFlZmIlNDBldGhlcmV1bS5tYWlsY2hhaW4uY29tIiwiZWZmZWN0IjoiQWxsb3ciLCJyZXNvdXJjZXMiOlsiKiJdLCJ0eXBlIjoiSG9sZGVyUG9saWN5In1dLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiTWFpbGNoYWluTWVzc2FnaW5nS2V5Q3JlZGVudGlhbCJdfV19fQ.g16IhL_pp8wL3YpZQUFlAdxquc6M_xyRWjLq_0q8RanAbpwJTMzdZ07B_yhk-ESF3dHREvNZJEB9IxkjGmbrDQ"`;

exports[`issuer mailchain-address-bob 2`] = `"{"aud":["did:mailchain:app-example%40mailchain.com"],"exp":1580602200,"iss":"did:mailchain:0x92d8f10248c6a3953cc3692a894655ad05d61efb%40ethereum.mailchain.com","nbf":1580601600,"nonce":"1234","vp":{"@context":["https://www.w3.org/2018/credentials/v1"],"type":["VerifiablePresentation"],"verifiableCredential":[{"@context":["https://www.w3.org/2018/credentials/v1"],"credentialSubject":[{"ownerOf":{"address":"0x92d8f10248c6a3953cc3692a894655ad05d61efb@ethereum.mailchain.com","type":"MailchainMessagingKey"}}],"issuanceDate":"2020-02-02T00:00:00.000Z","issuer":{"id":"did:mailchain:ethereum:0x92d8f10248c6a3953cc3692a894655ad05d61efb"},"proof":{"addressHex":"92d8f10248c6a3953cc3692a894655ad05d61efb","identityKeyHex":"e103bdf6fb97c97c126b492186a4d5b28f34f0671a5aacc974da3bde0be93e45a1c5","locale":"locale","messageVariant":"v1","messagingKeyHex":"e22e322f8740c60172111ac8eadcdda2512f90d06d0e503ef189979a159bece1e8","network":"","nonce":1,"protocol":"ethereum","signatureHex":"01020304","signatureMethod":"ethereum_personal_message","source":"MailchainRegistry","type":"Mailchain-Ed25519Signature2020"},"termsOfUse":[{"actions":["Authenticate","Join Meeting"],"assignee":"did:mailchain:app-example%40mailchain.com","assigner":"did:mailchain:0x92d8f10248c6a3953cc3692a894655ad05d61efb%40ethereum.mailchain.com","effect":"Allow","resources":["*"],"type":"HolderPolicy"}],"type":["VerifiableCredential","MailchainMessagingKeyCredential"]}]}}"`;
